### PR TITLE
docs(tracing): Introducing OpenTelemetry to distributed tracing

### PR DESCRIPTION
**Pull request by PaulRMellor**
Signed-off-by: prmellor <pmellor@redhat.com>
**Documentation**

Rewrite of [Introducing distributed tracing](https://strimzi.io/docs/operators/in-development/deploying.html#assembly-distributed-tracing-str) to introduce OpenTelemetry

- OpenTelemetry is prioritised
- Deprecation note for OpenTracing
- Content streamlined -- some assemblies removed to make the content easier find in the ToC, OpenTracing content that's replicated from OpenTracing documentation removed
- Rewrites in concepts
- Adds new procedures to switch tracing system
- Dependencies reduced to thos required
- Environment variables just show those required
- Includes OpenTelemetry equivalents in procedures
- Removes repeated steps for adding Kafka client dependencies. Instrumenting procedures now use a prereq for initlializing tracing for client

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [x] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: documentation/modules/tracing/proc-configuring-tracers-kafka-clients.adoc [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+[id='proc-configuring-tracers-kafka-clients-{context}']
+= Initializing tracing for Kafka clients
+
+[role="_abstract"]
+Initialize a tracer, then instrument your client applications for distributed tracing.
+You can instrument Kafka producer and consumer clients, and Kafka Streams API applications.
+You can initialize a tracer for OpenTracing or OpenTelemetry.
+
+Configure and initialize a Jaeger tracer using a set of xref:ref-tracing-environment-variables-{context}[tracing environment variables].
```
Comment (ppatierno) [inline documentation/modules/tracing/proc-configuring-tracers-kafka-clients.adoc:13]: let's remove "Jaeger" here

File: documentation/modules/tracing/proc-configuring-tracers-kafka-clients.adoc [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+<dependency>
+  <groupId>io.opentracing.contrib</groupId>
+  <artifactId>opentracing-kafka-client</artifactId>
+  <version>{OpenTracingKafkaClient}</version>
+</dependency>
+----
+
+. Define the configuration of the tracer using the xref:ref-tracing-environment-variables-{context}[tracing environment variables].
+
+. Create the tracer from the environment variables that you defined in step two:
```
Comment (ppatierno) [inline documentation/modules/tracing/proc-configuring-tracers-kafka-clients.adoc:58]: I don't get this sentence "from the environment variables" .... I would explain that you are using the OpenTracing or OpenTelemetry API to get a tracer which is initialized accordingly through the environment variables.

File: documentation/modules/tracing/proc-enabling-tracing-type.adoc [hunk 3]
```diff
@@ -0,0 +1,53 @@
+// Module included in the following assemblies:
+//
+// assembly-distributed tracing.adoc
+
+[id='proc-enabling-tracing-type-{context}']
+= Introducing a different OpenTelemetry tracing system
+
+[role="_abstract"]
+Instead of the default Jaeger system, you can specify other tracing systems that are supported by OpenTelemetry.
+You do this by adding the required artifacts to the library of the Kafka image provided with Strimzi.
```
Comment (ppatierno) [inline documentation/modules/tracing/proc-enabling-tracing-type.adoc:10]: "to the libs folder"? or just "adding the requered artifacts to the Kafka image ..."

File: documentation/modules/tracing/proc-enabling-tracing-type.adoc [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+metadata:
+  name: my-mm2-cluster
+spec:
+  #...
+  template:
+    connectContainer:
+      env:
+        - name: OTEL_SERVICE_NAME
+          value: my-zipkin-service
+        - name: OTEL_EXPORTER_JAEGER_ENDPOINT
```
Comment (ppatierno) [inline documentation/modules/tracing/proc-enabling-tracing-type.adoc:44]: this is `OTEL_EXPORTER_ZIPKIN_ENDPOINT` for zipkin

File: documentation/modules/tracing/proc-enabling-tracing-type.adoc [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+          value: http://zipkin-exporter-host-name:9411/api/v2/spans # <1>
+        - name: OTEL_TRACES_EXPORTER
+          value: zipkin # <2>
+  tracing:
+    type: opentelemetry
+#...
+----
+<1> Specifies the Zipkin endpoint to connect to.
+<2> The Zipkin exporter used for tracing.
```
Comment (ppatierno) [inline documentation/modules/tracing/proc-enabling-tracing-type.adoc:53]: I would not say "The Zipkin exporter ... " it sounds there are different types of Zipkin exporters and you have to specify one of them. It should just be "The exporter ... " and in this case it's Zipkin (not Jaeger).

File: documentation/modules/tracing/proc-configuring-tracers-kafka-clients.adoc [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+</dependency>
+----
++
+.Dependencies for OpenTracing
+[source,xml,subs="attributes+"]
+----
+<dependency>
+    <groupId>io.jaegertracing</groupId>
+    <artifactId>jaeger-client</artifactId>
+    <version>1.3.2</version>
```
Comment (ppatierno) [inline documentation/modules/tracing/proc-configuring-tracers-kafka-clients.adoc:47]: Shouldn't it be using `{JaegerClientVersion}`?

File: documentation/assemblies/tracing/assembly-distributed-tracing.adoc [hunk 7]
```diff
@@ -9,65 +9,71 @@ Distributed tracing allows you to track the progress of transactions between app
...
 
-Strimzi uses the OpenTracing and Jaeger Tracing projects to implement tracing.
-Tracing is enabled by setting the `.spec.tracing.type` property to `jaeger` in the component configuration.
+Tracing is not supported for Kafka brokers.
+You enable and configure tracing for these components using template configuration properties in their custom resources. 
+
+You specify a tracing type using the `tracing.type` property:
+
+* `opentelemetry` for OpenTelemetry with the Jaeger Exporter
+* `jaeger` for OpenTracing with Jaeger
```
Comment (scholzj) [inline documentation/assemblies/tracing/assembly-distributed-tracing.adoc:25]: The tracing is still enabled in `.spec.tracing.type`. Only the additional configuration options for the tracer are in the `.spec.template` section.

File: documentation/modules/tracing/proc-configuring-tracers-kafka-clients.adoc (hunk 6 above)
Comment (PaulRMellor) [reply documentation/modules/tracing/proc-configuring-tracers-kafka-clients.adoc:47]: Yes!

File: documentation/modules/tracing/proc-instrumenting-kafka-streams-with-tracers.adoc [hunk 8]
```diff
@@ -1,27 +1,56 @@
...
-// assembly-instrumenting-kafka-clients-tracers.adoc
+// assembly-distributed tracing.adoc
 
 [id='proc-instrumenting-kafka-streams-with-tracers-{context}']
 = Instrumenting Kafka Streams applications for tracing
 
-This section describes how to instrument Kafka Streams API applications for distributed tracing.
+[role="_abstract"]
+Instrument Kafka Streams API applications for distributed tracing.
+OpenTelemetry and OpenTracing instrumentation projects provide classes that support instrumentation of Kafka Streams.
```
Comment (ppatierno) [inline documentation/modules/tracing/proc-instrumenting-kafka-streams-with-tracers.adoc:10]: If we are referring to the `TracingKafkaClientSupplier` class, this is truer for OpenTracing but not for OpenTelemetry where the class is not provided out of box but we explain how to create it in the producer.

**Comment by ppatierno** [comment]
@PaulRMellor I realized the following ...

When explaining how to instrument producer and consumer we are talking about two ways: decorator and interceptor.
When we explain streams application we just talk about the decorator way through the "supplier" class which is a way to wrap the clients with tracing in a streams application. But ... it's also true you can use the interceptor way for streams application as well. It would use something like this 

```java
props.put(StreamsConfig.CONSUMER_PREFIX + ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
props.put(StreamsConfig.PRODUCER_PREFIX + ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingProducerInterceptor.class.getName());
```

File: documentation/assemblies/tracing/assembly-distributed-tracing.adoc (hunk 7 above)
Comment (PaulRMellor) [reply documentation/assemblies/tracing/assembly-distributed-tracing.adoc:25]: Changed as follows:

> You enable and configure tracing for these components through their custom resources. 
> You enable tracing by specifying a tracing type using the `spec.tracing.type` property:
> 
> * `opentelemetry` for OpenTelemetry with the Jaeger Exporter
> * `jaeger` for OpenTracing with Jaeger
> 
> You add tracing configuration using `spec.template` properties.

File: documentation/modules/tracing/proc-instrumenting-kafka-streams-with-tracers.adoc (hunk 8 above)
Comment (PaulRMellor) [reply documentation/modules/tracing/proc-instrumenting-kafka-streams-with-tracers.adoc:10]: Changed to:

> The OpenTracing instrumentation project provides a `TracingKafkaClientSupplier` class that supports instrumentation of Kafka Streams.
> For OpenTelemetry, you need to create a custom `TracingKafkaClientSupplier` class to provide the support.

**Comment by PaulRMellor** [comment]
> ```java
> props.put(StreamsConfig.CONSUMER_PREFIX + ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingConsumerInterceptor.class.getName());
> props.put(StreamsConfig.PRODUCER_PREFIX + ProducerConfig.INTERCEPTOR_CLASSES_CONFIG, TracingProducerInterceptor.class.getName());
> ```

@ppatierno Excellent. I've made the update in commit https://github.com/strimzi/strimzi-kafka-operator/pull/7403/commits/66c6c7b1ff31dfa08e80515c8290abc9607fb2fd.

I've assumed you might want to configure producer _or_ consumer. Maybe that should be _and_ for kafka streams.
I've also updated the span names section to say you can specify custom names when using the decorator pattern with Kafka Streams too.

**Comment by ppatierno** [comment]
@PaulRMellor to be honest I would use "producer AND consumer". Of course, what you are saying is possible in the sense that for a Kafka streams API application I could mix decorator and interceptor patterns.
For example, I can use interceptor for producer but then implements the supplier class for decorator returning a wrapped consumer but not a wrapped producer (because we are using tracing via interceptor for it).
Anyway, I would avoid suggesting something like that to the users. I don't see any value.
So let's try to be clear that you use interceptors for both producer AND consumer or you use the decorator (for both as well).

**Review by scholzj — APPROVED** [review]

**Review by ppatierno — APPROVED** [review]

**Pull request merged**